### ISSUE TITLE
Is there a better way to obtain all parameter values?

### ISSUE THREAD

**Issue by uduse** [deepmind/dm-haiku#151]
Say I want to compute the L2 loss of a given network, here's what I'm doing now:

```python
vecs = jax.tree_flatten(params)[0]
vecs = jax.tree_map(jnp.ravel, vecs)
vecs = jnp.concatenate(vecs)
loss = jnp.sum(vecs ** 2)
```

Is there a better way of doing so?

e.g., something like:
```python
loss = jnp.sum(hk.param_values(params))
```

**Comment by tomhennigan** [comment]
Hi @uduse , you can implement L2 loss given a params dict (or any jax pytree of arrays) as follows:

```python
def l2_loss(params):
  return 0.5 * sum(jnp.sum(jnp.square(p)) for p in jax.tree_leaves(params))
```

Typically you want to do this on a subset of parameters (e.g. ignore batch norm params). You can do that with Haiku's filter API before calling this API:

```python
l2_params = hk.data_structures.filter(lambda m, n, v: ('batchnorm' not in m), params)
loss = l2_loss(l2_params)
```

Our imagenet example uses l2 loss if you want to see this in a full example https://github.com/deepmind/dm-haiku/blob/master/examples/imagenet/train.py.

**Comment by samuela** [comment]
It's also worth noting that jax has an undocumented function `jax.flatten_util.ravel_pytree` that is similar to `tree_flatten` but actually ravels everything into one long array as well.
